README.md
# Buscacursos-uc plugin

Chrome web extension that lets you save more than one schedule in the [course-finder](http://buscacursos.uc.cl) PUC webpage.
See [extension](https://chrome.google.com/webstore/detail/buscacursosuc/fpekhaphlfpegpemjbcnekgbhkdkhgma) to install

***

Non-official extension, developed by [pdpino](https://github.com/pdpino).


docs/CHANGELOG.md
# Changelog

## Unreleased


## 0.4.0 - 2021-07-26
### Added
* Support changing semester (remove hardcoded value)
* Tooltip indicating semester must match with website
* Settings button (redirects to extension settings)


## 0.3.9 - 2021-07-24
### Changed
* Set 2021-2 as the default semester



## 0.3.8 - 2020-12-16
### Changed
* Set 2021-1 as the default semester



## 0.3.7 - 2020-08-10
### Added
* Empty placeholder in list
* Display text with form error

### Changed
* Overall style improvements
* Save `currentSemester` as field in each schedule
* Book icon

### Fixed
* Minor fixes in async messaging



## 0.3.6 - 2020-07-20
### Changed
* Set 2020-2 as the default semester

### Fixed
* Reload page after changes
* Remove unnecessary activeTab permission



## 0.3.5 - 2019-12-13
### Changed
* Set 2020-1 as the default semester


## 0.3.4 - 2019-07-01
### Changed
* Add current semester to HTML
* Add cursor pointer to buttons


## 0.3.3 - 2019-07-01
### Changed
* Set 2019-2 as the default semester


## 0.3.2 - 2018-30-11
### Changed
* Set 2019-1 as the default semester


## 0.3.1 - 2018-08-01
### Fixed
* Add expiration date to saved cookie, so it isn't saved as session cookie. This fixes a bug that caused deletion of the current schedule when closing the browser.

## 0.3.0 - 2018-07-23
### Added
* Button to override existing schedule with current one. A notification is shown
* Button to edit saved schedule's name

### Changed
* Wrap schedule name on button (not in whole item)

## 0.2 - 2018-07-22
### Changed
* Move logic functions that handle storage and cookies to background.js. Leave only html functions in popup.js

## 0.1 - 2018-07-20
### Added
* Basic functionalities:
  - Save current schedule using a name
  - Clear current schedule (remove all courses)
  - Load saved schedule
  - Remove saved schedules


js/popup.js
'use strict';

const saveScheduleButton = document.getElementById('save-schedule-button');
const clearScheduleButton = document.getElementById('clear-schedule-button');
const scheduleNameInput = document.getElementById('schedule-name-input');
const schedulesList = document.getElementById('schedules-list');
const schedulesListEmptyText = document.getElementById('schedules-list-empty-text');
const formErrorText = document.getElementById('form-error-text');
const currentSemesterText = document.getElementById('current-semester');
const prevSemesterButton = document.getElementById('prev-semester');
const nextSemesterButton = document.getElementById('next-semester');
const settingsButton = document.getElementById('settings-button');

/* Helper functions */
function removeElementById(id) {
  const element = document.getElementById(id);
  if (element) {
    element.parentNode.removeChild(element);
  }
}

function subscribeEnter(element, callback) {
  element.addEventListener('keyup', function(event) {
    if (event.key === 'Enter') {
      callback();
      event.preventDefault();
    }
  });
}

function sendNotification(message) {
  chrome.notifications.create('notification', {
    type: 'basic',
    iconUrl: '../img/ok.png',
    title: 'BuscacursosUC',
    message,
  }, function(notificationId) {
    setTimeout(function() {
      chrome.notifications.clear(notificationId);
    }, 1500);
  });
}

function showFormError(message) {
  formErrorText.classList.remove('disabled');
  formErrorText.textContent = message;
}

function hideFormError() {
  formErrorText.classList.add('disabled');
}


/* HTML elements functions */
function createButtonSelectSchedule(name) {
  const button = document.createElement('button');
  button.onclick = function() {
    chrome.runtime.sendMessage({ type: 'selectSchedule', name });
  };
  button.appendChild(document.createTextNode(name));
  button.setAttribute('class', 'name-schedule');
  button.setAttribute('title', 'Cargar horario');

  return button;
}

function createButtonChangeScheduleName(name, currentScheduleItem) {
  const button = document.createElement('button');
  button.setAttribute('class', 'change-schedule-name image-button');
  button.setAttribute('title', 'Editar nombre');
  button.onclick = function() {
    const editScheduleItem = document.createElement('li');
    const nameInput = document.createElement('input');
    nameInput.setAttribute('value', name);
    nameInput.setAttribute('type', 'text');
    nameInput.setAttribute('placeholder', 'Nombre');
    nameInput.setAttribute('class', 'name-schedule');

    const saveButton = document.createElement('button');
    saveButton.setAttribute('class', 'save-schedule-name image-button');
    saveButton.setAttribute('title', 'Guardar');
    saveButton.onclick = function() {
      const newName = nameInput.value;
      if (!newName) {
        console.log('USER ERROR: no name provided');
        return;
      }

      chrome.runtime.sendMessage({
        type: 'changeScheduleName',
        oldName: name,
        newName: newName,
      }, function(error) {
        if (error) return;
        const newItem = createScheduleItem(newName);
        schedulesList.replaceChild(newItem, editScheduleItem);
      });
    };
    subscribeEnter(nameInput, () => saveButton.click());

    editScheduleItem.appendChild(nameInput);
    editScheduleItem.appendChild(saveButton);

    schedulesList.replaceChild(editScheduleItem, currentScheduleItem);
  };

  return button;
}

function createButtonUpdateSchedule(name) {
  const button = document.createElement('button');
  button.onclick = function() {
    chrome.runtime.sendMessage({ type: 'updateSchedule', name }, function(error) {
      if (error) return; // TODO: change notification text??
      sendNotification(`'${name}' se sobrescribió con éxito`);
    });
  };
  button.setAttribute('class', 'update-schedule image-button');
  button.setAttribute('title', 'Sobreescribir horario');

  return button;
}

function createButtonDeleteSchedule(name) {
  const button = document.createElement('button');
  button.onclick = function() {
    chrome.runtime.sendMessage({ type: 'deleteSchedule', name }, function(error) {
      if (error) return;
      deleteScheduleItem(name);
    });
  };
  button.setAttribute('class', 'delete-schedule image-button');
  button.setAttribute('title', 'Eliminar horario');

  return button;
}

function createScheduleItem(name) {
  const scheduleItem = document.createElement('li');

  const selectScheduleButton = createButtonSelectSchedule(name);
  const changeScheduleNameButton = createButtonChangeScheduleName(name, scheduleItem);
  const updateScheduleButton = createButtonUpdateSchedule(name);
  const deleteScheduleButton = createButtonDeleteSchedule(name);

  scheduleItem.appendChild(selectScheduleButton);
  scheduleItem.appendChild(changeScheduleNameButton);
  scheduleItem.appendChild(updateScheduleButton);
  scheduleItem.appendChild(deleteScheduleButton);

  scheduleItem.setAttribute('id', name);

  return scheduleItem;
}

function renderScheduleItem(name) {
  const item = createScheduleItem(name);
  schedulesList.appendChild(item);
  renderEmptyList(false);
}

function renderEmptyList(empty) {
  if (empty) {
    schedulesList.classList.add('disabled');
    schedulesListEmptyText.classList.remove('disabled');
  } else {
    schedulesList.classList.remove('disabled');
    schedulesListEmptyText.classList.add('disabled');
  }
}

function deleteScheduleItem(name) {
  removeElementById(name);
  renderEmptyList(schedulesList.childElementCount === 0);
}

function renderSchedulesList(schedules) {
  if (schedules.length === 0) {
    renderEmptyList(true);
  } else {
    renderEmptyList(false);
    schedules.forEach(schedule => renderScheduleItem(schedule.name));
  }
}

function clearSchedulesList() {
  while(schedulesList.firstChild) {
    schedulesList.removeChild(schedulesList.firstChild);
  }
}

function loadDataAndUpdateUI(callback) {
  chrome.runtime.sendMessage({ type: 'loadData' }, function({ schedules, semester }) {
    clearSchedulesList();
    renderSchedulesList(schedules);
    currentSemesterText.textContent = semester;
    if (callback) {
      callback();
    }
  });
}

function _updateDataAndGoToSemesterURL(semester) {
  loadDataAndUpdateUI(function() {
    chrome.tabs.query({ active: true, currentWindow: true }, function(tabs) {
      chrome.tabs.update(tabs[0].id, {
        // NOTE: other url params will be lost
        // (typically from search queries)
        url: `https://buscacursos.uc.cl/?cxml_semestre=${semester}`,
      });
    });
  });
}

function clearScheduleInput() {
  scheduleNameInput.value = '';
}


/* Subscribe events */
saveScheduleButton.onclick = function() {
  const name = scheduleNameInput.value;
  if (!name) {
    console.log('USER ERROR: no name provided');
    showFormError('Provee un nombre');
    return;
  }
  chrome.runtime.sendMessage({ type: 'saveCurrentSchedule', name }, function(error) {
    if (error === 1) {
      showFormError('Horario vacío!');
      return;
    } else if (error === 2) {
      showFormError('Nombre ocupado!');
      return;
    }
    hideFormError();
    renderScheduleItem(name);
    clearScheduleInput();
  });
}

clearScheduleButton.onclick = function() {
  hideFormError();
  chrome.runtime.sendMessage({ type: 'clearCurrentSchedule' });
}

prevSemesterButton.onclick = function() {
  hideFormError();
  chrome.runtime.sendMessage({ type: 'prevSemester' }, _updateDataAndGoToSemesterURL);
}

nextSemesterButton.onclick = function() {
  hideFormError();
  chrome.runtime.sendMessage({ type: 'nextSemester' }, _updateDataAndGoToSemesterURL);
}

settingsButton.onclick = function() {
  chrome.runtime.openOptionsPage();
}

subscribeEnter(scheduleNameInput, () => saveScheduleButton.click());

/* First load (schedules and semester) */
loadDataAndUpdateUI();


js/options.js
'use strict'

// DEPRECATED (is not called from options.html)

const schedulesListContainer = document.getElementById('schedules-list-container');

function showSchedulesList() {
  chrome.storage.sync.get(['schedules'], function(result) {
    const schedules = result.schedules;
    if (schedules && schedules.length > 0) {
      const schedulesList = document.createElement('ul');
      schedules.forEach(function(schedule) {
        const item = document.createElement('li');
        item.appendChild(document.createTextNode(schedule.name));
        schedulesList.appendChild(item);
      });

      schedulesListContainer.appendChild(schedulesList);
    } else {
      const emptyText = 'No tienes horarios guardados';
      schedulesListContainer.appendChild(document.createTextNode(emptyText));
    }
  });
}

showSchedulesList();


js/background.js
'use strict';

chrome.runtime.onInstalled.addListener(function() {
  chrome.declarativeContent.onPageChanged.removeRules(undefined, function() {
    chrome.declarativeContent.onPageChanged.addRules([{
      conditions: [
        new chrome.declarativeContent.PageStateMatcher({
          pageUrl: { hostEquals: 'buscacursos.uc.cl' },
        }),
      ],
      actions: [
        new chrome.declarativeContent.ShowPageAction(),
      ],
    }]);
  });
});

/* Utils */
function getExpirationDate() {
  const nowDate = new Date();
  const nowSeconds = nowDate.getTime() / 1000;
  const monthsToAdd = 4;
  return nowSeconds + monthsToAdd*30*24*60*60;
}

function reloadPage() {
  chrome.tabs.query({ active: true, currentWindow: true }, function(tabs) {
    chrome.tabs.reload(tabs[0].id);
  });
}


/* Cookie wrappers */
function _getCurrentCookieDetail(semester) {
  return {
    url: 'http://buscacursos.uc.cl',
    name: `cursosuc-${semester}`,
  };
}

function removeCurrentCookie(semester, callback) {
  chrome.cookies.remove(_getCurrentCookieDetail(semester), callback);
}

function saveCurrentCookie(value, semester, callback) {
  chrome.cookies.set({
    ..._getCurrentCookieDetail(semester),
    expirationDate: getExpirationDate(),
    value,
  }, callback);
}

function getScheduleValue(semester, callback) {
  chrome.cookies.get(_getCurrentCookieDetail(semester), function(cookie) {
    if (!cookie || !cookie.value) {
      console.error('USER ERROR: no classes added to schedule');
      callback(null);
      return;
    }
    callback(cookie.value);
  });
}


/* Semester actions */
function _guessSemester() {
  const today = new Date();
  const year = today.getFullYear();
  const month = today.getMonth() + 1; // 0-based index
  const semester = (7 <= month && month <= 11) ? 2 : 1;
  return `${year}-${semester}`;
}

function _updateSemester(semester, callback) {
  chrome.storage.sync.set({ semester }, function() {
    callback(semester);
  });
}

function getSemester(callback) {
  chrome.storage.sync.get(['semester'], function(result) {
    let semester = result.semester;
    if (!semester) {
      semester = _guessSemester();
      _updateSemester(semester, callback);
    } else {
      callback(semester);
    }
  });
}

function prevSemester(callback) {
  getSemester(function(semester) {
    let [year, number] = semester.split('-').map(Number);
    if (number === 1) {
      number = 2;
      year -= 1;
    } else { // number === 2
      number = 1;
    }
    const newSemester = `${year}-${number}`;
    _updateSemester(newSemester, callback);
  });
}

function nextSemester(callback) {
  getSemester(function(semester) {
    let [year, number] = semester.split('-').map(Number);
    if (number === 2) {
      number = 1;
      year += 1;
    } else { // number === 1
      number = 2;
    }
    const newSemester = `${year}-${number}`;
    _updateSemester(newSemester, callback);
  });
}

/* Schedules basic actions */
function _loadSchedulesAndSemester(onlyThisSemester, callback) {
  chrome.storage.sync.get(['schedules', 'semester'], function(result) {
    const wasSemesterEmpty = !result.semester;
    const semester = result.semester || _guessSemester();
    let schedules = result.schedules || [];
    if (onlyThisSemester) {
      schedules = schedules.filter(sch => sch.semester === semester);
    }
    const output = {
      schedules,
      semester,
    };
    if (wasSemesterEmpty) {
      _updateSemester(semester, function() { callback(output) });
    } else {
      callback(output);
    }
  });
}

function loadThisSemesterSchedules(callback) { _loadSchedulesAndSemester(true, callback) };
function loadAllSchedules(callback) { _loadSchedulesAndSemester(false, callback) };

function saveSchedules(schedules, callback) {
  chrome.storage.sync.set({ schedules }, callback);
}


/* Schedules actions */
function deleteSchedule(name, callback) {
  loadAllSchedules(function({ schedules, semester }) {
    saveSchedules(schedules.filter(sch => sch.semester !== semester || sch.name !== name), callback);
  });
}

function selectSchedule(name) {
  loadAllSchedules(function({ schedules, semester }) {
    const schedule = schedules.find(sch => sch.semester === semester && sch.name === name);
    if (!schedule) {
      console.error(`INTERNAL ERROR: no schedule found with ${name} and semester ${semester}`);
      return;
    }
    saveCurrentCookie(schedule.value, semester, reloadPage);
  });
}

function saveCurrentSchedule(name, callback) {
  loadAllSchedules(function({ schedules, semester }) {
    getScheduleValue(semester, function(scheduleValue) {
      if (!scheduleValue) {
        callback(1);
        return;
      }
      if (schedules.find(sch => sch.semester === semester && sch.name === name)) {
        console.error('USER ERROR: name already taken, not overriding');
        callback(2);
        return;
      }
      schedules.push({
        name,
        value: scheduleValue,
        semester,
      });
      saveSchedules(schedules, callback);
    });
  });
}

function updateSchedule(name, callback) {
  loadAllSchedules(function({ schedules, semester }) {
    getScheduleValue(semester, function(scheduleValue) {
      if (!scheduleValue) {
        callback(1);
        return;
      }
      const index = schedules.findIndex(sch => sch.semester === semester && sch.name === name);
      if (index === -1) {
        console.error('INTERNAL ERROR: cannot update schedule, does not exist', name, semester);
        callback(2);
        return;
      }
      schedules[index].value = scheduleValue;
      saveSchedules(schedules, callback);
    });
  });
}

function clearCurrentSchedule() {
  getSemester(function(semester) {
    removeCurrentCookie(semester, reloadPage);
  });
}

function changeScheduleName(oldName, newName, callback) {
  if (oldName === newName) {
    callback();
    return;
  }

  loadAllSchedules(function({ schedules, semester }) {
    const index = schedules.findIndex(sch => sch.semester === semester && sch.name === oldName);
    if (index === -1) {
      console.error('INTERNAL ERROR: cannot change name, schedule does not exist', oldName, semester);
      callback(2);
      return;
    }
    if (schedules.find(sch => sch.semester === semester && sch.name === newName)) {
      console.error('USER ERROR: new name is taken');
      callback(3);
      return;
    }
    schedules[index].name = newName;
    saveSchedules(schedules, callback);
  });
}


/* Subscribe messages */
chrome.runtime.onMessage.addListener(
  function(request, sender, sendResponse) {
    switch (request.type) {
      case 'deleteSchedule':
        deleteSchedule(request.name, sendResponse);
        break;
      case 'selectSchedule':
        selectSchedule(request.name);
        break;
      case 'saveCurrentSchedule':
        saveCurrentSchedule(request.name, sendResponse);
        break;
      case 'updateSchedule':
        updateSchedule(request.name, sendResponse);
        break;
      case 'changeScheduleName':
        changeScheduleName(request.oldName, request.newName, sendResponse);
        break;
      case 'clearCurrentSchedule':
        clearCurrentSchedule();
        break;
      case 'loadData':
        loadThisSemesterSchedules(sendResponse);
        break;
      case 'prevSemester':
        prevSemester(sendResponse);
        break;
      case 'nextSemester':
        nextSemester(sendResponse);
        break;
      default:
        console.error('INTERNAL ERROR: no message listener for ', request);
    }
    return true;
  });
